### PR TITLE
New account settings design

### PR DESCRIPTION
**Pull request by nsharma123**

### CODE REVIEW

File: demo/index.html
Comment (latonv) [inline demo/index.html]: Q: Is this config info supposed to be in the public repo?

File: src/ia-account-settings.ts
Comment (latonv) [inline src/ia-account-settings.ts]: I think this condition can be simplified to `this.userData.screenname?.includes('\\')`. (The match result is only used for its truthiness, and both the `m` flag and `[...]` character class are superfluous here)

File: demo/index.html
Comment (iisa) [inline demo/index.html]: suggestion: let's not use a stringified json, let's use object since we are using lit's `render` function to append to page.

this goes for all of the properties that are stringified at entry. 👍

File: demo/index.html
Comment (iisa) [inline demo/index.html]: suggestion: let's remove this and add a placeholder to click on with demo

File: package.json
Comment (iisa) [inline package.json]: gentle reminder to push all the way to v1.0.0 when releasing

File: package.json
Comment (iisa) [inline package.json]: why remove lint? will we re-add before merge?

File: package.json
Comment (iisa) [inline package.json]: q: is web pack a direct dep? is it used by dev somehow?

File: src/assets/config.ts
Comment (iisa) [inline src/assets/config.ts]: suggestion: can we make an interface for this property value? that way we can easily share type

File: src/ia-account-settings.ts
Comment (latonv) [inline src/ia-account-settings.ts]: Blocker: the `await` needs to be applied directly to the `this.isEmailAvailable()` result (like with the screen name above). Right now this just negates the Promise object itself, so this condition will always be false.

File: src/ia-account-settings.ts
Comment (latonv) [inline src/ia-account-settings.ts]: Typo, I think. `emitExternalEvents`?

File: src/components/authentication-template.ts
Comment (iisa) [inline src/components/authentication-template.ts]: suggestion: make as button since it's clickable.

general rule, hit area is 30px x 30px

File: src/components/authentication-template.ts
Comment (iisa) [inline src/components/authentication-template.ts:195]: q: do we need a `submit` callback that does all checks before sending?

File: src/components/authentication-template.ts
Comment (iisa) [inline src/components/authentication-template.ts:182]: maybe this is its own form?

File: src/ia-account-settings.ts
Comment (latonv) [inline src/ia-account-settings.ts]: Q: What does `window.scrollTo()` without args do? I thought it required coordinates to be provided.

File: src/models.ts
Comment (iisa) [inline src/models.ts:47]: 👍

File: src/ia-account-settings.interface.ts
Comment (iisa) [inline src/ia-account-settings.interface.ts:43]: minor suggestion: add space to match file style

File: src/ia-account-settings.interface.ts
Comment (iisa) [inline src/ia-account-settings.interface.ts:78]: suggestion: remove jsdoc since we can inline the param and return types 🙂

File: src/ia-account-settings.interface.ts
Comment (iisa) [inline src/ia-account-settings.interface.ts]: q: are these just regular events or custom event? if custom, can we talk about its `e.detail` value/shape?

File: src/ia-account-settings.ts
Comment (latonv) [inline src/ia-account-settings.ts]: Q: Is the cast to `object` necessary here? Could `mailingLists` be given a more specific type?

File: src/ia-account-settings.ts
Comment (iisa) [inline src/ia-account-settings.ts]: I don't think we have to set this here as it receives from global scope?

**Review by iisa — CHANGES_REQUESTED** [review]
hello, midway through code review - I've added some comments, suggestions, questions for discourse 🙂 

q: can we make demo workable?

File: src/models.ts
Comment (latonv) [inline src/models.ts]: Q: Is it useful to have these properties typed to be exclusively the empty string literal?

File: src/services/backend-service.ts
Comment (latonv) [inline src/services/backend-service.ts]: Q: Do we want to remove this testing data?

File: src/services/backend-service.ts
Comment (latonv) [inline src/services/backend-service.ts]: Suggestion: since we're inside an `async` function anyway, stick to the async/await paradigm instead of switching to `then` callbacks here, for code simplicity.

File: src/services/backend-service.ts
Comment (latonv) [inline src/services/backend-service.ts]: No need to `await` the return value in an async function or a `then` callback, since you're really returning a Promise that resolves to that value anyway.

File: src/services/util.ts
Comment (latonv) [inline src/services/util.ts]: Suggestion: perhaps this function name could more accurately reflect that it does more than just `preventDefault`, to avoid confusion. Also, no need for this to be `async`.

File: test/ia-account-settings.test.ts
Comment (latonv) [inline test/ia-account-settings.test.ts]: Suggestion: uncapitalized `mailingLists`, since it is a normal value and not a class.

File: test/ia-account-settings.test.ts
Comment (latonv) [inline test/ia-account-settings.test.ts]: Q: I'm a bit confused about the type of `selectedMailingLists`. Where it's defined in the models, it looks like it should be a key-value dict. But here it's being set to an array. Could this be made more consistent?

**Review by latonv — CHANGES_REQUESTED** [review]
Nice job on this! Lots of solid work.

I pointed out what seems like a blocker in the email availability condition which should be addressed. Otherwise, just some other suggestions & questions throughout to help improve code clarity and type accuracy.

File: demo/index.html
Comment (nsharma123) [reply demo/index.html]: No, we are passing at account/index.php file petabox

File: package.json
Comment (nsharma123) [reply package.json]: Not used. removed

File: src/components/authentication-template.ts
Comment (nsharma123) [reply src/components/authentication-template.ts:182]: we don't need separate form as we both (third-party-auth and ia) sharing single form without noise.

File: src/ia-account-settings.interface.ts
Comment (nsharma123) [reply src/ia-account-settings.interface.ts]: these are regular events.

File: src/services/backend-service.ts
Comment (nsharma123) [reply src/services/backend-service.ts]: We no need to remove to as these are for /demo/ host (on localserver)

File: src/assets/config.ts
Comment (nsharma123) [reply src/assets/config.ts]: yes, I added an interface type in  models.ts

File: src/ia-account-settings.ts
Comment (nsharma123) [reply src/ia-account-settings.ts]: Yes, I have changed these almost all property structure in way to use specific type.

File: test/ia-account-settings.test.ts
Comment (nsharma123) [reply test/ia-account-settings.test.ts]: I have changed the property types

File: src/components/authentication-template.ts
Comment (nsharma123) [reply src/components/authentication-template.ts:195]: no we don't need here.

File: src/services/backend-service.ts
Comment (nsharma123) [reply src/services/backend-service.ts]: removed asycn/await

File: src/ia-account-settings.interface.ts
Comment (nsharma123) [reply src/ia-account-settings.interface.ts:78]: I wanted to consolidated at single place for all the functions I am using so added all the jsdocs here only.

File: src/components/authentication-template.ts
Comment (latonv) [inline src/components/authentication-template.ts]: Q: Trimming whitespace off a _password_ seems very unusual to me. Is there a requirement for this? Shouldn't we use the exact password they entered without alteration?

File: src/ia-account-settings.ts
Comment (latonv) [inline src/ia-account-settings.ts]: (As with the comment above, not sure why we need to trim passwords)

File: src/ia-account-settings.ts
Comment (latonv) [inline src/ia-account-settings.ts]: Suggestion: This looks like it should be a button, rather than a void anchor. (We can still style it to look like a link).

Also: Is the `window.scrollTo()` necessary? Without x/y coordinates, isn't that a no-op?

File: src/ia-account-settings.ts
Comment (latonv) [inline src/ia-account-settings.ts]: Suggestion: Move the `for="borrow-history"` to the first label (the one that identifies what the checkbox is for).

File: src/ia-account-settings.ts
Comment (latonv) [inline src/ia-account-settings.ts]: Suggestion: Correct the doc comments for these properties

File: src/ia-account-settings.ts
Comment (latonv) [inline src/ia-account-settings.ts]: Q: What does adding the `loanHistoryFlag` to the field name do?

File: src/ia-account-settings.ts
Comment (latonv) [inline src/ia-account-settings.ts:354]: Q: I'm still a bit confused by the type of `selectedMailingLists`. It's initialized to an empty object `{}` above, but here it's having array methods called on it. Can we double-check this?

File: src/services/util.ts
Comment (latonv) [inline src/services/util.ts]: Suggestion: lowercase `number` type? (And then you shouldn't need the type cast below)

**Review by latonv — COMMENTED** [review]
Thanks for the updates! Just had some more suggestions & questions on a few things.

File: src/ia-account-settings.ts
Comment (nsharma123) [reply src/ia-account-settings.ts]: The first label is heading and the second is that identify the checkbox. so we are good here.

File: src/ia-account-settings.ts
Comment (nsharma123) [reply src/ia-account-settings.ts]: Intented to remove around space from all user inputs. BTW removed for password. Thanks!

File: src/ia-account-settings.ts
Comment (latonv) [reply src/ia-account-settings.ts]: Ah okay, I misunderstood the meaning of the checkbox. Sounds good 👍

File: src/assets/eye.-open.ts
Comment (latonv) [inline src/assets/eye.-open.ts]: Typo: I think the filename has an unwanted extra `.` in it (`src/assets/eye.-open.ts`)

File: src/ia-account-settings.ts
Comment (latonv) [inline src/ia-account-settings.ts]: A couple more JSDocs to update the descriptions for

File: src/ia-account-settings.ts
Comment (latonv) [inline src/ia-account-settings.ts:783]: Minor suggestion: Might be nice to just add a comment nearby noting what the entities `&#10003;` and `&#10006;` are, so that other maintainers don't have to look it up.

File: package.json
Comment (latonv) [inline package.json]: Minor detail, but could we update the repo description/README?

**Review by latonv — APPROVED** [review]
LGTM, just a few final small suggestions.

Also, just a reminder to add linting back to the `test` script before merge, as @iisa noted.

**Pull request merged**
